### PR TITLE
sway: improve systemdIntegration with dbus

### PR DESCRIPTION
**Pull request by genofire**
used for pinentry-gnome3 e.g.

### Checklist

- [ ] Change is backwards compatible.

- [ ] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

----
adopted from archlinux:
https://github.com/archlinux/svntogit-community/blob/packages/sway/trunk/50-systemd-user.conf

### CODE REVIEW

File: modules/services/window-managers/i3-sway/sway.nix [hunk 1]
```diff
@@ -312,7 +312,9 @@ let
     ''
   else
     "") + "\n" + (if cfg.systemdIntegration then ''
-      exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+      exec "systemctl --user import-environment DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+      exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+      exec "systemctl --user start sway-session.target"
```
Comment (berbiche) [inline modules/services/window-managers/i3-sway/sway.nix:317]: I don't think Sway guarantees the order of execution so you should just split the command over multiple lines with `\\n` and add a semicolon between them.

**Review by berbiche — CHANGES_REQUESTED** [review]
Sorry for missing this PR and thanks for the contribution!

File: modules/services/window-managers/i3-sway/sway.nix (hunk 1 above)
Comment (DPDmancul) [reply modules/services/window-managers/i3-sway/sway.nix:317]: In the official wiki (https://github.com/swaywm/sway/wiki#gtk-applications-take-20-seconds-to-start) they use different `exec`s

Comment (berbiche) [reply modules/services/window-managers/i3-sway/sway.nix:317]: Fair enough.

**Comment by berbiche** [comment]
The logs have expired and I cannot re-run the workflow.

All the tests have to be updated since the generated configuration has changed.

``` console
$ nix eval --raw --impure --expr 'builtins.concatStringsSep "\n" (builtins.attrNames (import ./tests {}).run)' | grep sway | xargs -I{} nix-shell -I nixpkgs=channel:nixos-21.11 --pure tests -A run.{}
```

I fixed the tests and took the liberty to add `XDG_CURRENT_DESKTOP` to the imported variables.
This is useful with `xdg-desktop-portal` and `xdg-desktop-portal-{gtk,wlr}`.

**Comment by berbiche** [comment]
Well, I cannot push (or force-push) the fix to your branch for whatever reason.

Here's the patch (save it and apply it with `git apply`):

``` patch
commit 5eb23e1dcefbbf1b292ce3a1049205d0ddb2f178
Author: Nicolas Berbiche <nicolas@normie.dev>
Date:   Mon Jan 31 23:06:22 2022 -0500

    sway: fix tests

diff --git a/modules/services/window-managers/i3-sway/sway.nix b/modules/services/window-managers/i3-sway/sway.nix
index 58e7b427..d8a7195a 100644
--- a/modules/services/window-managers/i3-sway/sway.nix
+++ b/modules/services/window-managers/i3-sway/sway.nix
@@ -312,10 +312,10 @@ let
     ''
   else
     "") + (concatStringsSep "\n" ((optional cfg.systemdIntegration ''
-      exec "systemctl --user import-environment DISPLAY WAYLAND_DISPLAY SWAYSOCK"
-      exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY SWAYSOCK"
-      exec "systemctl --user start sway-session.target"
-    '') ++ (optional (!cfg.xwayland) "xwayland disable")
+      exec "systemctl --user import-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+      exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+      exec "systemctl --user start sway-session.target"'')
+      ++ (optional (!cfg.xwayland) "xwayland disable")
       ++ [ cfg.extraConfig ])));
 
   defaultSwayPackage = pkgs.sway.override {
diff --git a/tests/modules/services/window-managers/sway/sway-bar-focused-colors.conf b/tests/modules/services/window-managers/sway/sway-bar-focused-colors.conf
index d55c9200..d8d92d97 100644
--- a/tests/modules/services/window-managers/sway/sway-bar-focused-colors.conf
+++ b/tests/modules/services/window-managers/sway/sway-bar-focused-colors.conf
@@ -110,4 +110,6 @@ bar {
   
 }
 
-exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+exec "systemctl --user import-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "systemctl --user start sway-session.target"
diff --git a/tests/modules/services/window-managers/sway/sway-bindkeys-to-code.conf b/tests/modules/services/window-managers/sway/sway-bindkeys-to-code.conf
index 97c72710..3fe780d5 100644
--- a/tests/modules/services/window-managers/sway/sway-bindkeys-to-code.conf
+++ b/tests/modules/services/window-managers/sway/sway-bindkeys-to-code.conf
@@ -110,4 +110,6 @@ bar {
   
 }
 
-exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+exec "systemctl --user import-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "systemctl --user start sway-session.target"
diff --git a/tests/modules/services/window-managers/sway/sway-default.conf b/tests/modules/services/window-managers/sway/sway-default.conf
index 6ef61a7a..e8fe38e8 100644
--- a/tests/modules/services/window-managers/sway/sway-default.conf
+++ b/tests/modules/services/window-managers/sway/sway-default.conf
@@ -110,4 +110,6 @@ bar {
   
 }
 
-exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+exec "systemctl --user import-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "systemctl --user start sway-session.target"
diff --git a/tests/modules/services/window-managers/sway/sway-followmouse-expected.conf b/tests/modules/services/window-managers/sway/sway-followmouse-expected.conf
index 7fa4da07..a8b5e6c7 100644
--- a/tests/modules/services/window-managers/sway/sway-followmouse-expected.conf
+++ b/tests/modules/services/window-managers/sway/sway-followmouse-expected.conf
@@ -83,4 +83,6 @@ bindsym k resize shrink height 10 px
 bindsym l resize grow width 10 px
 }
 
-exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+exec "systemctl --user import-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "systemctl --user start sway-session.target"
diff --git a/tests/modules/services/window-managers/sway/sway-followmouse-legacy-expected.conf b/tests/modules/services/window-managers/sway/sway-followmouse-legacy-expected.conf
index c470a7c4..8996ae2e 100644
--- a/tests/modules/services/window-managers/sway/sway-followmouse-legacy-expected.conf
+++ b/tests/modules/services/window-managers/sway/sway-followmouse-legacy-expected.conf
@@ -83,4 +83,6 @@ bindsym k resize shrink height 10 px
 bindsym l resize grow width 10 px
 }
 
-exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+exec "systemctl --user import-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "systemctl --user start sway-session.target"
diff --git a/tests/modules/services/window-managers/sway/sway-modules.conf b/tests/modules/services/window-managers/sway/sway-modules.conf
index c0339d71..b029f2f7 100644
--- a/tests/modules/services/window-managers/sway/sway-modules.conf
+++ b/tests/modules/services/window-managers/sway/sway-modules.conf
@@ -122,4 +122,6 @@ bar {
   
 }
 
-exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+exec "systemctl --user import-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "systemctl --user start sway-session.target"
diff --git a/tests/modules/services/window-managers/sway/sway-null-package.conf b/tests/modules/services/window-managers/sway/sway-null-package.conf
index 6ef61a7a..e8fe38e8 100644
--- a/tests/modules/services/window-managers/sway/sway-null-package.conf
+++ b/tests/modules/services/window-managers/sway/sway-null-package.conf
@@ -110,4 +110,6 @@ bar {
   
 }
 
-exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+exec "systemctl --user import-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "systemctl --user start sway-session.target"
diff --git a/tests/modules/services/window-managers/sway/sway-workspace-default-expected.conf b/tests/modules/services/window-managers/sway/sway-workspace-default-expected.conf
index b5c92fa6..3b740479 100644
--- a/tests/modules/services/window-managers/sway/sway-workspace-default-expected.conf
+++ b/tests/modules/services/window-managers/sway/sway-workspace-default-expected.conf
@@ -109,4 +109,6 @@ bar {
   
 }
 
-exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+exec "systemctl --user import-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "systemctl --user start sway-session.target"
diff --git a/tests/modules/services/window-managers/sway/sway-workspace-output-expected.conf b/tests/modules/services/window-managers/sway/sway-workspace-output-expected.conf
index bacb27f8..a209e8ce 100644
--- a/tests/modules/services/window-managers/sway/sway-workspace-output-expected.conf
+++ b/tests/modules/services/window-managers/sway/sway-workspace-output-expected.conf
@@ -114,4 +114,6 @@ workspace "1" output eDP
 workspace "ABC" output DP
 workspace "3: Test" output HDMI
 workspace "!"§$%&/(){}[]=?\*#<>-_.:,;²³" output DVI
-exec "systemctl --user import-environment; systemctl --user start sway-session.target"
+exec "systemctl --user import-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "hash dbus-update-activation-environment 2>/dev/null && dbus-update-activation-environment XDG_CURRENT_DESKTOP DISPLAY WAYLAND_DISPLAY SWAYSOCK"
+exec "systemctl --user start sway-session.target"
```

**Pull request closed without merging**